### PR TITLE
[+] Search - Developers can configure in which fields the search will be executed.

### PR DESCRIPTION
**Pull request by lucasscariot**
## Pull Request checklist:

- [x] Write an explicit title for the Pull Request
- [x] Write changes made in the CHANGELOG.md
- [ ] Create automatic tests
- [ ] No automatic tests failures
- [x] Test manually the implemented changes
- [x] Review my own code (indentation, syntax, style, simplicity, readability)
- [x] Wonder if you can improve the existing code

### CODE REVIEW

**Comment by lucasscariot** [comment]
[(2) Advanced/Custom search in Rails](https://trello.com/c/bwA1Pg4s/1744-2-advanced-custom-search-in-rails)

File: app/services/forest_liana/resources_getter.rb [hunk 1]
```diff
@@ -5,7 +5,9 @@ def initialize(resource, params)
       @params = params
       @count_needs_includes = false
       @field_names_requested = field_names_requested
-
+      @current_collection = ForestLiana.apimap.find do |collection|
+        collection.name.to_s == @resource.table_name
+      end
```
Comment (arnaudbesnier) [inline app/services/forest_liana/resources_getter.rb]: You duplicate code here (see same code at line 64 of the same file)

File: app/services/forest_liana/has_many_getter.rb [hunk 2]
```diff
@@ -5,6 +5,9 @@ def initialize(resource, association, params)
       @association = association
       @params = params
       @field_names_requested = field_names_requested
+      @current_collection = ForestLiana.apimap.find do |collection|
+        collection.name.to_s == association_table_name
+      end
```
Comment (arnaudbesnier) [inline app/services/forest_liana/has_many_getter.rb]: You should create a method in the parent class `BaseGetter` for this. The implementation would be at a single place for the `ResourcesGetter` and the `HasManyGetter`.

File: app/services/forest_liana/search_query_builder.rb [hunk 3]
```diff
@@ -50,23 +51,25 @@ def search_param
...
         @resource.columns.each_with_index do |column, index|
           column_name = format_column_name(@resource.table_name, column.name)
-          if column.name == 'id'
-            if column.type == :integer
-              value = @params[:search].to_i
-              conditions << "#{@resource.table_name}.id = #{value}" if value > 0
-            elsif REGEX_UUID.match(@params[:search])
-              conditions << "#{@resource.table_name}.id =
-                '#{@params[:search]}'"
+          if (@collection.search_fields && @collection.search_fields.index(column.name)) || !@collection.search_fields
```
Comment (arnaudbesnier) [inline app/services/forest_liana/search_query_builder.rb]:
```ruby
if @collection.search_fields.nil? || (@collection.search_fields && @collection.search_fields.include?(column.name))
```
The "search_fields" condition is almost always `true` so lets make the code more readable and optimise the code execution here.

File: app/services/forest_liana/search_query_builder.rb [hunk 4]
```diff
@@ -81,15 +84,24 @@ def search_param
 
         if (@params['searchExtended'].to_i == 1)
           SchemaUtils.one_associations(@resource).map(&:name).each do |association|
+            if @collection.search_fields
+              association_search = @collection.search_fields.map do |field|
+                if field.include? association.to_s
```
Comment (arnaudbesnier) [inline app/services/forest_liana/search_query_builder.rb]: For a more secure code you should check that:
```ruby
if field.include? '.' && field.split('.')[0] == association.to_s
  field.split('.')[1]
end
```

Comment (arnaudbesnier) [reply app/services/forest_liana/search_query_builder.rb]: I am also pretty sure that `association_search` looks like an array with `nil` values and should not for better performances.
screen shot 2017-09-06 at 09 22 37

File: app/services/forest_liana/search_query_builder.rb [hunk 5]
```diff
@@ -81,15 +84,24 @@ def search_param
...
               resource = @resource.reflect_on_association(association.to_sym)
               resource.klass.columns.each do |column|
                 if !(column.respond_to?(:array) && column.array) &&
                   (column.type == :string || column.type == :text)
                   column_name = format_column_name(resource.table_name,
                     column.name)
-                  conditions << "LOWER(#{column_name}) LIKE " +
-                    "'%#{@params[:search].downcase}%'"
+                    if (association_search && association_search.include?(column.name)) || !association_search
+                    conditions << "LOWER(#{column_name}) LIKE " +
```
Comment (arnaudbesnier) [inline app/services/forest_liana/search_query_builder.rb]:
```ruby
if @collection.search_fields.nil? || (association_search && association_search.include?(column.name))
```

**Comment by arnaudbesnier** [comment]
Code example for tests:
```ruby
search_fields %w{ delivery_address customer.address }
# or
search_fields ['delivery_address', 'customer.address']
```

File: app/services/forest_liana/base_getter.rb [hunk 6]
```diff
@@ -1,5 +1,10 @@
 module ForestLiana
   class BaseGetter
+    def get_current_collection(table_name)
+      ForestLiana.apimap.find do |collection|
+        collection.name.to_s == table_name
+      end
+    end
```
Comment (arnaudbesnier) [inline app/services/forest_liana/base_getter.rb]: A newline after the method declaration would be more conventional in the current code.

File: app/services/forest_liana/resources_getter.rb [hunk 7]
```diff
@@ -5,7 +5,7 @@ def initialize(resource, params)
       @params = params
       @count_needs_includes = false
       @field_names_requested = field_names_requested
-
+      @current_collection = get_current_collection(@resource.table_name)
```
Comment (arnaudbesnier) [inline app/services/forest_liana/resources_getter.rb]: Some duplicated code remains at line 62.

File: app/services/forest_liana/search_query_builder.rb [hunk 8]
```diff
@@ -50,7 +51,9 @@ def search_param
 
         @resource.columns.each_with_index do |column, index|
           column_name = format_column_name(@resource.table_name, column.name)
-          if column.name == 'id'
+          if (@collection.search_fields && !@collection.search_fields.index(column.name))
```
Comment (arnaudbesnier) [inline app/services/forest_liana/search_query_builder.rb]: `index` vs `include?` => include is really much more readable, more "ruby style"

File: app/services/forest_liana/search_query_builder.rb [hunk 9]
```diff
@@ -81,15 +84,24 @@ def search_param
 
         if (@params['searchExtended'].to_i == 1)
           SchemaUtils.one_associations(@resource).map(&:name).each do |association|
+            if @collection.search_fields
+              association_search = @collection.search_fields.map do |field|
+                if field.include?('.') && field.split('.')[0] == association.to_s
+                  field.split('.')[1]
+                end
+              end
```
Comment (arnaudbesnier) [inline app/services/forest_liana/search_query_builder.rb]: here, again, I am sure that the resulting array `association_search` contains `nil` values that should not be here for performance and code quality (a `compact` call should do the job)

**Pull request merged**